pytroll_collectors/s3stalker.py

```diff
@@ -23,37 +19,133 @@
+        self.close()
+
+    def run(self):
+        """Start the s3-stalker daemon/runner in a thread."""
+        logger.info("Starting up... ")
+        self._setup_and_start_communication()
+
+        first_run = True
+        last_fetch_time = None
+        while self.loop:
+            self._set_timedelta(last_fetch_time, first_run)
+
+            last_fetch_time = get_last_fetch()
+            logger.info("Last fetch time...: %s", str(last_fetch_time))
+            first_run = False
+
+            self._process_messages()
+
+            logger.debug("Waiting %d seconds", self._wait_seconds)
+            time.sleep(max(self._wait_seconds, 0))
+
+    def _set_timedelta(self, last_fetch_time, first_run):
+        self._timedelta = self._get_timedelta(last_fetch_time, is_first_run=first_run)
+
+    def _process_messages(self):
+        """Go through all messages in list and publish them one after the other."""
+        messages = create_messages_for_recent_files(self.bucket, self.config, self._timedelta)
+        for message in messages:
+            logger.info("Publishing %s", str(message))
+            self.publisher.send(str(message))
+
+    def _get_timedelta(self, last_fetch_time, is_first_run):
+        """Get the seconds for the time window to search for (new) files."""
+        if is_first_run:
+            logger.info('Create messages with urls for most recent files only')
+            logger.info('On start up we consider files with age up to %d seconds from now',
+                        self.startup_timedelta_seconds)
+            return {'seconds': self.startup_timedelta_seconds}
+
+        seconds_back = self._get_seconds_back_to_search(last_fetch_time)
+        return {'seconds': seconds_back}
+
+    def _get_seconds_back_to_search(self, last_fetch_time):
+        """Update the time to look back considering also the modification time of the last file."""
+        if last_fetch_time is None:
+            return self._wait_seconds
+
+        start_time = datetime.utcnow()
+        start_time = start_time.replace(tzinfo=timezone.utc)
+        seconds_to_last_file = (start_time - last_fetch_time).total_seconds()
+        return max(self._wait_seconds, seconds_to_last_file)
+
+    def close(self):
+        """Shutdown the S3Stalker runner."""
+        logger.info('Terminating the S3 Stalker daemon/runner.')
+        self.loop = False
+        if self.publisher:
+            self.publisher.stop()
+
+
 @contextmanager
 def sleeper(duration):
     """Make sure the block takes at least *duration* seconds."""
     start_time = datetime.utcnow()
     yield
     end_time = datetime.utcnow()
     waiting_time = duration - (end_time - start_time).total_seconds()
+    logger.debug('waiting time: %f', waiting_time)
     time.sleep(max(waiting_time, 0))
 
 
 class DatetimeHolder:
     """Holder for the last_fetch datetime."""
 
-    last_fetch = datetime.now(tz.UTC) - timedelta(hours=12)
+    last_fetch = datetime.utcnow() - timedelta(hours=12)
```

Comment:
If I read correctly, this class is used as a globla variable, right? Could the `last_fetch` time be attribute of the `S3StalkerRunner` class? That way there would be no need for `get/set_last_fetch()` functions and `get_last_files()` could be a method without `*args` and `**kwargs`.

Reply:
Help wanted! @mraspaud Did this code

Reply:
We can move that to another PR, as I commented just before I was somehow confused what already existed and what didn't.